### PR TITLE
R2RDump - Dump AvailableTypes section contents

### PR DESCRIPTION
**Pull request by acmyu**

### CODE REVIEW

**Comment by acmyu** [comment]
The only new commit in this pr is "Move R2RReader constructor to separate functions, parse READYTORUN_SECTION_AVAILABLE_TYPES"

File: src/tools/r2rdump/R2RReader.cs [hunk 1]
```diff
@@ -65,130 +71,177 @@ public unsafe R2RReader(string filename)
...
+                    method.RuntimeFunctions.Add(new RuntimeFunction(runtimeFunctionId, startRva, endRva, unwindRva, method));
+                    runtimeFunctionId++;
                 }
+                while (runtimeFunctionId < isEntryPoint.Length && !isEntryPoint[runtimeFunctionId]);
+            }
+        }
+
+        private void ParseAvailableTypes()
+        {
+            R2RSection instMethodEntryPointSection = R2RHeader.Sections[R2RSection.SectionType.READYTORUN_SECTION_AVAILABLE_TYPES];
```
Comment (zamont) [inline src/tools/r2rdump/R2RReader.cs]: Can you rename this to something more appropriate like `availableTypesSection`?

File: src/tools/r2rdump/R2RReader.cs [hunk 2]
```diff
@@ -65,130 +71,177 @@ public unsafe R2RReader(string filename)
...
+            int instMethodEntryPointsOffset = GetOffset(instMethodEntryPointSection.RelativeVirtualAddress);
```
Comment (zamont) [inline src/tools/r2rdump/R2RReader.cs]: Rename too?

File: src/tools/r2rdump/R2RReader.cs [hunk 3]
```diff
@@ -65,130 +71,177 @@ public unsafe R2RReader(string filename)
...
+            NativeHashtable.AllEntriesEnumerator allEntriesEnum = instMethodEntryPoints.EnumerateAllEntries();
+            NativeParser curParser = allEntriesEnum.GetNext();
+            while (!curParser.IsNull())
+            {
+                uint rid = curParser.GetUnsigned();
+                rid = rid >> 1;
+                TypeDefinitionHandle typeDefHandle = MetadataTokens.TypeDefinitionHandle((int)rid);
+
+                TypeDefinition typeDef = _mdReader.GetTypeDefinition(typeDefHandle);
+                Tuple<string, string> pair = new Tuple<string, string>(_mdReader.GetString(typeDef.Namespace), _mdReader.GetString(typeDef.Name));
```
Comment (zamont) [inline src/tools/r2rdump/R2RReader.cs]: Two comments:

1) How come you don't combine the namespace and type name right here? Do you plan to use them separately later on?
2) You might be able to use the _mdReader for when you're getting the name of methods.

File: src/tools/r2rdump/R2RReader.cs (hunk 1 above)
Comment (acmyu) [reply src/tools/r2rdump/R2RReader.cs]: oops, I forgot to rename it

File: src/tools/r2rdump/R2RReader.cs (hunk 3 above)
Comment (acmyu) [reply src/tools/r2rdump/R2RReader.cs]: I don't use them separately right now since it's only printing it, but would it be useful to have them separate?

**Pull request merged**